src/posix/platform/multicast_routing.cpp

```diff
@@ -0,0 +1,564 @@
+}
+
+void MulticastRoutingManager::Remove(const Ip6::Address &aAddress)
+{
+    otError error = OT_ERROR_NONE;
+
+    assert(mListenerSet.find(aAddress) != mListenerSet.end());
+    mListenerSet.erase(aAddress);
+
+    VerifyOrExit(IsEnabled());
+
+    RemoveInboundMulticastForwardingCache(aAddress);
+
+exit:
+    otLogResultPlat(error, "MulticastRoutingManager: %s: %s", __FUNCTION__, aAddress.ToString().AsCString());
+}
+
+void MulticastRoutingManager::UpdateFdSet(fd_set &aReadFdSet, int &aMaxFd) const
+{
+    VerifyOrExit(IsEnabled());
+
+    FD_SET(mMulticastRouterSock, &aReadFdSet);
+    aMaxFd = std::max(aMaxFd, mMulticastRouterSock);
+
+exit:
+    return;
+}
+
+void MulticastRoutingManager::Process(const fd_set &aReadFdSet)
+{
+    VerifyOrExit(IsEnabled());
+
+    if (FD_ISSET(mMulticastRouterSock, &aReadFdSet))
+    {
+        ProcessMulticastRouterMessages();
+    }
+
+exit:
+    return;
+}
+
+otError MulticastRoutingManager::InitMulticastRouterSock(void)
+{
+    otError             error = OT_ERROR_NONE;
+    int                 one   = 1;
+    struct icmp6_filter filter;
+    struct mif6ctl      mif6ctl;
+
+    VerifyOrExit(mMulticastRouterSock == -1);
+
+    // Create a Multicast Routing socket
+    mMulticastRouterSock = socket(AF_INET6, SOCK_RAW, IPPROTO_ICMPV6);
+    VerifyOrExit(mMulticastRouterSock != -1, error = OT_ERROR_FAILED);
+
+    // Enable Multicast Forwarding in Kernel
+    VerifyOrExit(0 == setsockopt(mMulticastRouterSock, IPPROTO_IPV6, MRT6_INIT, &one, sizeof(one)),
+                 error = OT_ERROR_FAILED);
+
+    // Filter all ICMPv6 messages
+    ICMP6_FILTER_SETBLOCKALL(&filter);
+    VerifyOrExit(0 == setsockopt(mMulticastRouterSock, IPPROTO_ICMPV6, ICMP6_FILTER, (void *)&filter, sizeof(filter)),
+                 error = OT_ERROR_FAILED);
+
+    memset(&mif6ctl, 0, sizeof(mif6ctl));
+    mif6ctl.mif6c_flags     = 0;
+    mif6ctl.vifc_threshold  = 1;
+    mif6ctl.vifc_rate_limit = 0;
+
+    // Add Thread network interface to MIF
+    mif6ctl.mif6c_mifi = kMifIndexThread;
+    mif6ctl.mif6c_pifi = if_nametoindex(gNetifName);
+    VerifyOrExit(mif6ctl.mif6c_pifi > 0, error = OT_ERROR_FAILED);
+    VerifyOrExit(0 == setsockopt(mMulticastRouterSock, IPPROTO_IPV6, MRT6_ADD_MIF, &mif6ctl, sizeof(mif6ctl)),
+                 error = OT_ERROR_FAILED);
+
+    // Add Backbone network interface to MIF
+    mif6ctl.mif6c_mifi = kMifIndexBackbone;
+    mif6ctl.mif6c_pifi = if_nametoindex(gBackboneNetifName);
+    VerifyOrExit(mif6ctl.mif6c_pifi > 0, error = OT_ERROR_FAILED);
+    VerifyOrExit(0 == setsockopt(mMulticastRouterSock, IPPROTO_IPV6, MRT6_ADD_MIF, &mif6ctl, sizeof(mif6ctl)),
+                 error = OT_ERROR_FAILED);
+
+exit:
+    if (error != OT_ERROR_NONE)
+    {
+        FinalizeMulticastRouterSock();
+    }
+
+    return error;
+}
+
+void MulticastRoutingManager::FinalizeMulticastRouterSock(void)
+{
+    VerifyOrExit(mMulticastRouterSock != -1);
```

Comment:
The current implementation is `{ return mMulticastRouterSock >= 0; }`, but it could change in the future. 

`FinalizeMulticastRouterSock` focus on the closing the socket if it's opened. I prefer `VerifyOrExit(mMulticastRouterSock != -1);` since it clearly states the condition of this method.

Reply:
revised